front/lib/swr/data_source_documents.ts

```diff
@@ -0,0 +1,58 @@
+import type { DataSourceViewType, LightWorkspaceType } from "@dust-tt/types";
+import type {
+  PatchDataSourceWithNameDocumentRequestBody,
+  PostDataSourceWithNameDocumentRequestBody,
+} from "@dust-tt/types";
+import type { SWRMutationConfiguration } from "swr/mutation";
+import useSWRMutation from "swr/mutation";
+
+import { fetcherWithBody } from "@app/lib/swr/swr";
+import type { PostDocumentResponseBody } from "@app/pages/api/w/[wId]/spaces/[spaceId]/data_sources/[dsId]/documents";
+import type { PatchDocumentResponseBody } from "@app/pages/api/w/[wId]/spaces/[spaceId]/data_sources/[dsId]/documents/[documentId]";
+
+async function sendPatchRequest(
+  url: string,
+  {
+    arg,
+  }: {
+    arg: {
+      documentBody: PatchDataSourceWithNameDocumentRequestBody;
+    };
+  }
+) {
+  const res = await fetcherWithBody([url, arg.documentBody, "PATCH"]);
+  return res;
+}
+
+export function useUpdateDataSourceDocumentMutation(
+  owner: LightWorkspaceType,
+  dataSourceView: DataSourceViewType,
+  documentName: string,
+  options?: SWRMutationConfiguration<PatchDocumentResponseBody, Error, string>
+) {
+  const patchUrl = `/api/w/${owner.sId}/spaces/${dataSourceView.spaceId}/data_sources/${dataSourceView.dataSource.sId}/documents/${documentName}`;
+  return useSWRMutation(patchUrl, sendPatchRequest, options);
+}
+
+async function sendPostRequest(
+  url: string,
+  {
+    arg,
+  }: {
+    arg: {
+      documentBody: PostDataSourceWithNameDocumentRequestBody;
+    };
+  }
+) {
+  const res = await fetcherWithBody([url, arg.documentBody, "POST"]);
+  return res;
+}
+
+export function useCreateDataSourceDocumentMutation(
+  owner: LightWorkspaceType,
+  dataSourceView: DataSourceViewType,
+  options?: SWRMutationConfiguration<PostDocumentResponseBody, Error, string>
+) {
+  const createUrl = `/api/w/${owner.sId}/spaces/${dataSourceView.spaceId}/data_sources/${dataSourceView.dataSource.sId}/documents`;
```

Comment:
weird to mutate a url used for POST

Reply:
This seems fine to me and is used in the SWR doc as well https://swr.vercel.app/docs/mutation#useswrmutation

Reply:
Ah ok, they do it because GET `/api/user` return the user (and in our case /api/w/${owner.sId}/spaces/${dataSourceView.spaceId}/data_sources/${dataSourceView.dataSource.sId}/documents return the list of docs).

Looking at the backend code, here, you'll only mutate call with the path (not any query strings args) and it seems that we call this endpoint with params. For this, we have `mutateRegardlessOfQueryParams()` that is returned by `useSWRWithDefaults()`. Another reason imho to use our current pattern for fetching + mutating.

Let me know if you want me to show you IRL.